### ISSUE TITLE
Release Notes

### ISSUE THREAD

**Issue by tabatkins** [tabatkins/bikeshed#1773]
Version 2.0.0
=============

* Feature: Added darkmode styles!

	This is the reason for the version bump, as if you were adding your own styles to your document, they probably aren't darkmode-aware and will look bad when the readers are in darkmode. If you're just using Bikeshed's own styles, tho, everything will work exactly as before.

* Feature: [Added new conditional-inclusion feature](https://tabatkins.github.io/bikeshed/#conditional), allowing you to remove sections of a spec (typically in boilerplate) based on the document's Status. If you have more things you want to base a conditional inclusion off of, let me know!

	(Currently, the processing order of this happens after all the 'datablocks' (`pre` with special class that triggers Bikeshed handling) happens, so even if those are inside of a conditional that gets excluded, they'll still get processed. But everything else happens after conditional stuff is removed.)

* Bugfix: [The rest of IDL's built-in types are now autolinked.](https://github.com/tabatkins/bikeshed/commit/428ace32b90c02fc39c872be53a3d916f51350de)

**Comment by tabatkins** [comment]
Version 2.0.0
=============

* Feature: Added darkmode styles!

	This is the reason for the version bump, as if you were adding your own styles to your document, they probably aren't darkmode-aware and will look bad when the readers are in darkmode. If you're just using Bikeshed's own styles, tho, everything will work exactly as before.

* Feature: [Added new conditional-inclusion feature](https://tabatkins.github.io/bikeshed/#conditional), allowing you to remove sections of a spec (typically in boilerplate) based on the document's Status. If you have more things you want to base a conditional inclusion off of, let me know!

	(Currently, the processing order of this happens after all the 'datablocks' (`pre` with special class that triggers Bikeshed handling) happens, so even if those are inside of a conditional that gets excluded, they'll still get processed. But everything else happens after conditional stuff is removed.)

* Bugfix: [The rest of IDL's built-in types are now autolinked.](https://github.com/tabatkins/bikeshed/commit/428ace32b90c02fc39c872be53a3d916f51350de)

**Comment by tabatkins** [comment]
Version 2.1.0
=============

* Feature enhancement: [Conditional inclusion](https://tabatkins.github.io/bikeshed/#conditional) can now conditionalize on the presence of a [custom boilerplate element](https://tabatkins.github.io/bikeshed/#overriding-boilerplate) (something with `boilerplate=foo` attribute) or the presence of a [custom text macro](https://tabatkins.github.io/bikeshed/#metadata-text-macro), so you can include a chunk of related HTML only when the thing it's trying to refer to actually exists!
* Feature: Autolink conjugations (the thing that lets you write `[=applying=]` when the original definition was just `<dfn>apply</dfn>` now apply to the *first* word of a phrase as well as the last. This will *usually* catch the verb in a multi-word phrase, like "apply the widget", so you can write `[=applying the widget=]` and have it Just Work. (But all the conjugations apply, so noun-headed terms like "widget controlled by the user" can be autolinked with `[=widgets controlled by the user=]` now.)
* Feature: Biblio autolinks (like `[[FOO]]`) can now be switched into "inline" display mode, where rather than displaying as "[FOO]" and linking to the bibliography, they display as the actual document's title and link straight to them. (They still get added to the index.) You can [turn it on individually](https://tabatkins.github.io/bikeshed/#biblio) with `[[FOO inline]]`, or set the default for all biblios in the spec with [Default Biblio Display](https://tabatkins.github.io/bikeshed/#metadata-default-biblio-display) (and switch them back to "index" mode with `[[FOO index]]`).

* Bugfix: Better darkmode styles! Various fixes thruout, including to certain boilerplates.
* Bugfix: bumped the versions for several dependencies to the most recent form, just for general dependency-hygiene reasons
* Bugfix: If you used `Editor Term` to rename what the spec-metadata section called "Editors", various small side-effects that should specially trigger for the Editor lines... wouldn't. They do now.

**Comment by tabatkins** [comment]
Version 2.3.0
=============

Whoops, didn't write release notes for the 2.2 versions! Oh well!

* Feature: Global `-a` flag to force Bikeshed to emit ASCII-only in all its messages.
* Feature: When using `WPT Path Prefix` with a *lot* of tests missing, don't spam the full list out, just log a warning that there are a bunch.
* Feature: Biblio autolinks can now take the `obsolete` keyword to allow an intentionally-obsolete reference (otherwise Bikeshed complains and automatically switches your ref to what obsoleted it).
* Feature: Finally got daily test updates working, so we run with the most modern set of specs rather than ones that are obsolete by months.

* Bugfix: Stop ignoring certain test types from WPT, so any test can be linked.
* Bugfix: Track what data files exist locally manually, rather than relying on the list matching what the remote data last contained.
* Bugfix: Grab the correct set of boilerplates for the specified Status.

* Misc: Lots and *lots* of fixes to make Bikeshed lint-clean.

**Comment by tabatkins** [comment]
Version 2.3.1
=============

* A bunch of dependency bumping
* More importantly, [Automatic Image Size Detection](https://tabatkins.github.io/bikeshed/#img-size), care of @frivoal! If your `<img>` elements don't already have width/height attributes on them, Bikeshed will do its best to auto-detect what the sizes are, and insert them into your markup automatically. This is a small but nice perf improvement for specs, avoiding layout-jiggle as images load in.

Version 2.4.0
=============

* Added the NOTE-ED, NOTE-WD, and NOTE-FPWD statuses, for handling the new somewhat-more-official "Note track" for W3C documents, so you can have boilerplates with appropriate verbiage to indicate it's a Note-track document. (REGISTRY-* statuses will come in a bit, when the updates to W3C Process finish up enough to allow publishing such a document.)
* Minor fixes to boilerplates and styles. Shouldn't be anything you notice, but if you're tracking diffs in your generated output you'll see some chatter.
* Another banger from @frivoal, [`<wpt>` elements have some new features]:
	* `<wpt hidden>` processes the element but omits it from the output (useful for leveled specs that need to include all the tests in their folder so they can benefit from `WPT Prefix`, but don't actually have a place to put the tests for up/down-level versions)
	* `<wpt title>` lets you specify a descriptive paragraph for the tests in the block, formatting it just above the test list in the output
	* Multiple adjacent `<wpt>` elements will now be merged in the output
	* Best of all, rendered `<wpt>` elements are now `<details>` elements, and the new `WPT Display: closed` renders them closed by default (the previous `inline` value and its new alias `open` render them open by default, looking approximately the same as they previously did). The default styling for a closed `<wpt>` is a tiny out-of-the-way "TESTS", suitable for displaying even in live, published specs. Play with it for now, consider adding it to your `defaults.include` file. I plan to flip this to be the default in a near-future version.

**Comment by tabatkins** [comment]
Version 2.4.1
=============

* Several small boilerplate tweaks to some groups
* <wpt title> attribute is now parsed as Bikeshed source

**Comment by tabatkins** [comment]
Version 2.4.2
=============

* A few boilerplate updates, as always
* Properly strip the internal-use attributes from `<pre class=include-code>` elements from the output document, so they won't cause your document to fail HTML validation.


Version 2.4.3
=============

* Boilerplates again.
* A couple of widlparser dependency updates

**Comment by tabatkins** [comment]
lol been skipping release notes for a while!

Version 2.4.4
==========
* boilerplates again

Version 2.4.5
==========
* boilerplates
* removed some ancient (and now incorrect) lines from the linking-defaults file

Version 2.4.6
==========
* accidentally did the previous release on Windows and picked up Windows line endings

Version 2.4.7
===========
* fixed the url link-defaults was pulled from during update!
* boilerplates
* bumped a lot of dependencies
* fixed <https://github.com/tabatkins/bikeshed/commit/3d037a4f7270476ea76fa4fd2d6ec406e33750aa>, which caused text to get dropped in some circumstances

Version 3.0.0
==========
* New major version! Nothing exciting, I just pushed a breaking change for security reasons; you now have to explicitly opt-in, from the command line, to allowing Bikeshed to grab data from outside the input file's directory, or running arbitrary code via some extensions features.
* biblio entries and links now use `cite`, such semantic <https://github.com/tabatkins/bikeshed/commit/e4a654b463a551cb611500b18bd5f777e2103132>
* boilerplates

Version 3.0.1
==========
* boilerplates
* actually lint the 3.0.0 code that restricted use of files outside the source directory

Version 3.0.2
===========
* boilerplates
* some correctness fiddling to the 3.0.0 code

Version 3.0.3
==========
* ??? I did literally nothing, weird.

Version 3.0.4
==========
* boilerplates (jeez I'm getting tired of this)

Version 3.0.5
==========
* boilerplates

Version 3.0.6
==========
* boilerplates

Version 3.1.0
===========
* NO MORE BOILERPLATES UPDATES. Bikeshed now stores and versions its boilerplates over at <https://github.com/tabatkins/bikeshed-boilerplate>, and brings in updates via `bikeshed update` like any other data files. 

Version 3.2.0
========
* fix a lot of lint errors from 3.1.0
* propdef tables now link their "Applies To" header <https://github.com/tabatkins/bikeshed/commit/980486aa7d3fc11d8711118c5938064b4e72a2cb>
* Issues lists (from `bikeshed issues-list`) can now have intro text <https://github.com/tabatkins/bikeshed/commit/c2f0309eeade8c20550a4cbd08c79349f8fe7db1>
* Format links to GH issues better <https://github.com/tabatkins/bikeshed/commit/8bb502091c0fa6a322c3b402ca297bf382715ecd>
* Added AOM as a group and megagroup, plus some new statuses for them

Version 3.3.0
===========
Big new WPT features - inline test results!
* `<wpt>` blocks now contain pass/fail information in the main browsers! <https://github.com/tabatkins/bikeshed/commit/f07e1bf21590ce19405c02e483aa4520b4877ef4>
* `wpt` lines indicate when a test file contains multiple tests, and how many
* If there's a WPT path prefix (or I can infer one), add the corresponding http://wpt.fyi link in the spec-metadata section up top
* also add pass/fail data for all spec's tests combined to the spec-metadata section

And some other random stuff:
* Stop doing the weird "implicitly call .format() on messages" and just use f-strings or explicit .format() everywhere <https://github.com/tabatkins/bikeshed/commit/368df1cf2bb640348227ff65d56b6f17da1830c2>
* Update the dockerfile <https://github.com/tabatkins/bikeshed/commit/24054d8e3fc4e2470137535ce54b9cd11d56242c>
* propdef tables now link their "Canonical Order" header <https://github.com/tabatkins/bikeshed/commit/ed3425a59ec5f546f79b35df2d6e2d84fd3bf94a>
* `bikeshed template` now outputs a Status that actually works, so you can use its output immediately, as intended.

Version 3.3.1
==========
* adapt to the Process 2021 changes that got sprung on me without announcement

Version 3.3.2
==========
* I'd accidentally been omitting the wpt data files from the distributed version. They'll be as out of data by default as all the other built-in data files, but they should at least *exist*. Fixed now. <https://github.com/tabatkins/bikeshed/commit/bbc364800e88c2472ca75ad7c5c8c3a09505a4b4>

Version 3.3.3
==========
* remove some now-unused W3C NOTE statuses <https://github.com/tabatkins/bikeshed/commit/6cf917509c94f769dea7d498f7fbe093c41ca809>
* Only set the LOGO macro if Logo metadata is actually set <https://github.com/tabatkins/bikeshed/commit/31a2f03f4befb89325bebb9c5f25a84dce5fa119>
* Allow conditional logic to look at non-custom metadata too; combined with previous, you can now add a default logo if a spec doesn't set their own logo. <https://github.com/tabatkins/bikeshed/commit/31a2f03f4befb89325bebb9c5f25a84dce5fa119>

Version 3.3.4
==========
* CanIUse blocks are now closed by default, like MDN blocks <https://github.com/tabatkins/bikeshed/commit/be0e7d0e15e69d1ce41dc87393a3a1d728d3f7fc>
* Fix update link for boilerplate files <https://github.com/tabatkins/bikeshed/commit/a892fb19de40ebeaa698cff58dd55ed4f777dcc6>
* Actually clean up old boilerplate files that no longer exist <https://github.com/tabatkins/bikeshed/commit/cace53f8cdd3fe04ceef2801e15a5f7fe8706630>

Version 3.4.0
==========
* Add pngwg to W3C groups (I need to move this data out of this repo, too.....) <https://github.com/tabatkins/bikeshed/commit/c9308f39fdde8a1360d92b3d108747fd91533198>
* New "direct" biblio style that still formats the link text as the shortname (to match the biblio entry), but skips the biblio and links straight to the document. <https://github.com/tabatkins/bikeshed/commit/1f1661843554a272b4a14c35637caf3f2e16fa8a>
* Documented the W3C's spec-prod GH action, which makes it easy to auto-publish with Bikeshed
* Disallow merged Privacy/Security sections, per new W3C rules <https://github.com/tabatkins/bikeshed/commit/893894c7db84fc2f6610539350c69563ba4495df>
* When you've just got a few headings missing their ID, output them individually so they get line numbers <https://github.com/tabatkins/bikeshed/commit/c29faa6f6b9210d0cf9d6f03202e009bb83257db>
* Document how to set "html metadata" (attributes on `html`, `meta`/`link`/`style`/`script` elements) <https://github.com/tabatkins/bikeshed/commit/8eaccd9f127bedc5566dcac0c27aa196601adad7>
* Detect and scream when you're likely hitting the dumb years-old argparser bug about spaces in the values of unknown flags <https://github.com/tabatkins/bikeshed/commit/beb1f2893fe20963f26d9285c8eed88b99b927c4>
* Strip the common WS prefix from fenced code blocks, same as I do for `pre` elements <https://github.com/tabatkins/bikeshed/commit/53a81004f02671f93a75e16f7b9ad85a24fd8474>
* Add a new debug command to output all the metadata passed to the document, from all the possible sources <https://github.com/tabatkins/bikeshed/commit/cd66b7b4622d0e5e396ef95308d80c7f8f74577e>

**Comment by tabatkins** [comment]
Version 3.4.1
==========
* Previously, Bikeshed would helpfully *not* spam you with all the missing tests that matched your WPT Path Prefix if there were more than a handful, since there were legit reasons to have a bunch of untracked tests (tests not for your level, maybe you're in the middle of annotating, etc). Now that we have `<wpt hidden>` to do a well-targeted silencing when required, that's no longer a problem, so let'er rip. <https://github.com/tabatkins/bikeshed/commit/f965859122fd22d04fb9094bf7c5b00b5ba48ca6>
* If you've only got empty/hidden `wpt` elements, Bikeshed no longer adds the JS, spec-metadata entries, etc. Style still comes thru so the blocks render, at least. <https://github.com/tabatkins/bikeshed/commit/4d354f1e58e5760fb0f2db49a08bd005b0bdc6db>
* The legacy processor Bikeshed replaced handled links to properties and descriptors identically; 'font-family' would gladly link to the property *or* the @font-face descriptor, depending on what was in the spec. That never worked well in the first place (the Fonts spec would randomly link to each), and while Bikeshed could at least *detect* the issue and complain about it, it's annoying to deal with when properties are *usually* what you wanted to link to. It's gotten worse as we've added more descriptors whose names collide with properties, like 'width' or 'aspect-ratio'.
	
	So, propdesc autolinks without an explicit for value are now always treated as property autolinks; `'width'` will *always* link to the 'width' property, and never ask if you meant the 'width' MQ or CQ descriptor. You can still use the syntax with a for value, like `'@media/width'`, to link to descriptors. This might cause some new link failures, but more likely it'll cause some existing link failures to go away and link correctly. <https://github.com/tabatkins/bikeshed/commit/f9584d2c0f419a220c9d60f9998c3ed8d846496f>

**Comment by tabatkins** [comment]
# Version 3.5.0

* Fixed the automatic indent-removal for pre blocks in some cases <https://github.com/tabatkins/bikeshed/commit/bd9a15d0ee3f32ebd7e3d6a476c04482ec7fe15c>
* Added the `[:http header:]` autolink syntax
* Added the "permission" definition type <https://github.com/tabatkins/bikeshed/commit/f092c3aaae972eeb294877fb0bd5cca45537230d>
* Trust anchor-block entries, even if the spec you're linking to is marked as obsoleted in SpecRef <https://github.com/tabatkins/bikeshed/commit/b1c6bfae8395782aa4ef1290724acc755171e41a>
* Allow `div` around `dt`/`dd`, including markdown-generated ones, since that's valid in HTML. Previously it would often *kinda* work *by accident* if you contorted your markup in a certain way, but it was real bad. <https://github.com/tabatkins/bikeshed/commit/790e00888925ef861f2f4b1246c5f17736f7c090>
* Lots of adventures in beginning add typing information to Bikeshed. Long way to go yet.
* Properly autolink the type signatures in argumentdef blocks. <https://github.com/tabatkins/bikeshed/commit/ac0d42042966d025b5b8a547289665305eabfa53>
* Added a few W3C groups and statuses

# Version 3.5.1
* Added the fedidcg group

# Version 3.5.2
* Made section autolinks also trigger a biblio entry, which I believe is the last place that should be able to do so. Also localized the code that tracks this so all the entry points do the exact same thing; this means some of your biblio entries might change casing since they're consistent across all callsites now. There's still some mismatch between what specs I know about via biblio data and what specs I know about from definition data, which I know will cause at least a few instances where "the same" biblio entry shows up twice under slightly different names, but that's a separate issue that won't be fixed until I move Bikeshed's definitions entirely to Reffy. <https://github.com/tabatkins/bikeshed/commit/fdd5936bf8dbfedd59b6f5837ea5f8a0e0db0752>

**Comment by tabatkins** [comment]
# Version 3.6

* `<div assert>` now works, to get the same auto-generated ID as `<assert>` does, but for block-level. (`<assert>` turns itself into a `<span>` in the output.) (Thanks, @dj2!)
* Fixed some new duplication of normative/informative refs caused by the section autolinks fixes from 3.5.2.

# Version 3.7
* Can now build multi-file specs directly from an (uncompressed) TAR file! This is worthless *locally*, but you can use this to send multi-file specs to the API server. (Thanks, @kainino0x!)

**Comment by tabatkins** [comment]
Been a while since my last batch of release notes!

Version 3.7.1
=============

* WPT tests mentioned multipled times in a spec now properly only count once for the test counts (https://github.com/tabatkins/bikeshed/pull/2304)
* WPT summaries use the proper Stable channel browser logos (https://github.com/tabatkins/bikeshed/pull/2296)
* Started some serious mypy-ification

Version 3.7.2
=============

* Fix a bug I caused while mypy-ing.
* More mypy-ification.

Version 3.7.3
=============

* Fix a minor bug from an argument being removed in py 3.9.
* More mypy-ification. (Seeing a trend? This continues until October.)

Version 3.7.4
=============

* Just mypy stuff. Unsure why I cut this version, honestly?

Version 3.7.5
=============

* Fix the merging of adjacent MDN blocks.
* Fix caniuse and MDN darkmode styles.
* Made the markdown em/strong replacers work better. They're still not *right*, but they're now *less wrong*.
* `Complain About: broken-links yes` now has a timeout so it can't hang indefinitely.
* Fixed a cache accidentally getting shared across documents, continually growing when you generate a bunch of specs (like `bikeshed test` does...)
* Fixed how heading data is recognized from the data source, to better handle how ReSpec formats its markup.
* FINISHED MYPY-IFICATION

Version 3.7.6
=============

* Properly add some new non-py files introduced by 3.7.5 to the release data.

Version 3.7.7
=============

* Fixed a spot where if you force-generate (so `die()` doesn't stop the program) it would end up with inconsistent state and cause a Python error anyway.

Version 3.8.0
=============

* Added simple def tables, formatted like propdef/etc but arbitrary keys
* Added `bikeshed update --force-manifest` that forces it to do a manifest update even if the remote's date indicates your local data should be more recent.

Version 3.8.1
=============

* Fix anchor recognition, which was broken in 3.7.5 by the heading fix.

Version 3.8.2
=============

* Added a missing `from __future__ import annotations` that was causing a crash in older Python versions.
* Only warn about missing WPT tests under your specified prefix once, rather than twice.

Version 3.9.0
=============

* Properly flag if your anchor-block refs have a local URL (which breaks other stuff anyway).
* Fix a long-standing issue where section autolinks could generate a similar or duplicate biblio entry to what was already generated by definition autolinks.
* Obsolete biblio links are now LINKERROR, not FATAL.
* Moved the self-links from notes/issues/examples to avoid being cut off by their 'overflow' value.
* HTTP Header autolinks now use the correct typography per Infra.
* Added `Complain About: mixed-indents yes` to catch when you mix tabs and spaces in bad ways.

**Comment by tabatkins** [comment]
New big batch of release notes:

Version 3.10.0
============

* Added explicit support for "Logical Property Group" in propdef blocks

Version 3.10.1
============

* Biblio links now have the reference's title as their `title` (unless they're `inline`, since those already have the title as their text)
* Better parsing for Markdown URL syntax. (Think it's still not perfect, but substantially improved wrt parsing the url portion.)
* Fixed the `bikeshed echidna --just-tar` option to actually echo the TAR file, rather than whatever broken nonsense it had ended up doing.
* Fixed a bug with negative ranges in the CSS `<length [-100px, 100px]>` syntax, and just switched to using hyphen-minus rather than the fancy math minus.

Version 3.11.0
============

* Switched Bikeshed's ref database to using the W3C's WebRef service, rather than plinss's Shepherd service. Lots of commits related to this.
* Note markers (the "Note:" bit) now have class=marker, as the docs said they should.

Version 3.11.1
============
* Reworked the "unused var" error-checking so I could output line numbers, to help find the error.

Version 3.11.2 & 3.11.3
===========
* Bikeshed moved official ownership! Now a project under the Specification Editors CG in the W3C, at <https://github.com/speced/bikeshed>.

Version 3.11.4
============
* After some poking at WebRef to get the data complete, Bikeshed no longer relies on Shepherd at all.
* Put a *really cute* little progress bar on both the manual update info, and test running.

Version 3.11.5
============
* Fixed the "mixed indentation" lint to properly allow tabs followed by spaces. (Spoilers: it's still broken at this point.)
* Added the first part of my new bespoke HTML parser.
* Fiddled with the dfn panel generation/positioning for better a11y.

Version 3.11.6
============
* Fixed the "mixed indentation" lint correctly. 
* Try to infer the indent from source analysis, rather than just looking at the first indented line.

Version 3.11.7
============
* Fiddle with ID generation for links that dfn panels need to link to, to make them more likely to be unique.

Version 3.11.8
============
* Fix how I package bikeshed for release, so I'll stop forgetting to include new .css and .js files when I add them.

Version 3.11.9
============
* Fixed a filepath.

Version 3.11.10
=============
* Removed a good chunk of the ARIA stuff I tried to add, since it technically violated the ARIA spec (even tho I think it was right). Grumble, grumble.

Version 3.11.11
=============
* Unify a lot of the CanIUse/MDN panel code, particularly positioning. Switch them from doing some funky abspos-with-static-position stuff to being positioned purely with JS, since they can trivially overlap otherwise.
* Specs marked as "fork" in WebRef are no longer included in the db.
* A bunch of fiddling with the update code to untangle the old mess of transforms needed for mixing the Shepherd and WebRef sources. Since it's all WebRef now I can make it *much* simpler.

Version 3.11.12
=============
* Accidental rerelease, lol

Version 3.11.13
=============
* Include .bsfont files in the build too. Surprised this is the first time someone's run into this issue!

Version 3.11.14
=============
* Preserve the start-tag's attributes when handling all the various datablocks (propdef, railroad, include, etc).
* Move the semver file into the bikeshed folder so it's actually installed by pip. Turns out the top-level files aren't!

Version 3.11.15
=============
* Nothing, just didn't notice the semver got bumped again while I was fiddling with the release script.

Version 3.11.16
=============
* Since 3.11.14 started using the new bespoke HTML parser, it can accidentally surface some HTML mistakes that were previously silently accepted. The error messages weren't showing the right line tho - now fixed.

**Comment by tabatkins** [comment]
# Version 3.11.17

Turns out using just a *little* of a new HTML parser causes problems when my old hacky line-based code is, well, doing things line-by-line, and thus only feeding the parser part of a tag when a tag is split across multiple lines. So now I just use the HTML parser for the entire pre-Markdown parsing pipeline. (Then Markdown runs, generates some HTML, and feeds its result straight to the lxml parser. Both of these will be rewritten in time as well.)

It looks like this doesn't cause any issues with the testsuite, and hopefully should resolve the spurious fatal errors that 3.11.16 was throwing. Lmk if you see any issues!

**Comment by domenic** [comment]
Thanks Tab for the quick response and the postmortem. I'm really excited about work on tracking build messages, and also only making changes as PRs on GitHub. The way in which ~half of the commits on https://github.com/speced/bikeshed/commits/main have red Xs next to them has made me very nervous about using Bikeshed in the past, and moving to a PRs-with-CI-checks model would be a huge boost to my confidence in the software. Especially once those CI checks include build messages.

**Comment by domenic** [comment]
Oh. That's disappointing. I wish you would treat main with the same care as you treat releases, and used CI to validate your changes, like modern software best practice requires (e.g. on the Chromium project we both contribute to).

Now that Bikeshed is part of SpecEd, is there a way to ask "SpecEd" for that sort of working mode?

**Comment by tabatkins** [comment]
> I wish you would treat main with the same care as you treat releases, and used CI to validate your changes, like modern software best practice requires (e.g. on the Chromium project we both contribute to).

Why? This isn't a rhetorical question - what is the reasoning behind you wanting this?

Chromium is a massive multi-contributor project; it needs a core that is stable/correct for all the contributors to work from. Bikeshed is largely a single-contributor (me) project, and I don't believe my working mode is what's stopping more contributions. Bureaucracy exists for a good reason, but it's also a significant drag on its host; this is more than worthwhile when it's cutting thru the N^2 growth of personal connections between contributors, but not when that N is very low (again, approximately 1 here).

The group of people that *do* need stability in reference to Bikeshed are, currently, just the users, and they're served well by my release process. *You* are also in that group; as WHATWG uses the API server, which now follows releases rather than main, you're exposed solely to what I do in releases. What happens in main is irrelevant in practice for you.

> Now that Bikeshed is part of SpecEd, is there a way to ask "SpecEd" for that sort of working mode?

No. I moved Bikeshed to SpecEd for continuity reasons; my getting hit by a bus will no longer mean that the development has to fork to continue. I did not move it to enable non-contributors to impose currently-unneeded bureaucracy on the project.

**Comment by domenic** [comment]
It's sad to me that you see proper software engineering practices (which are used by teams of all sizes, including one) as bureaucracy. If there's such an ideological gap at this point, I don't think we'll be able to make further progress through discussion, at least not until more incidents of this sort (or others we've seen throughout Bikeshed's history) accumulate.

**Comment by dlaliberte** [comment]
I wonder if there is just a misunderstanding going on here.  Given my limited knowledge of git, I may be missing something, but a few things are clear to me, and as one of the contributors of a small fraction of changes to Bikeshed, I may have a few useful insights.

One thing we should be able to agree on is that the error status of commits should be considered totally irrelevant as long as the last commit before squashing and merging is "green".  The commits are just intermediate states that are steps along the path toward a final PR submission, which is true both for main and any branches.   In fact, I find the extra steps of staging changes, then committing changes, and then syncing changes, almost entirely annoyingly extraneous.

Doing independent work on branches separate from origin/main, such that once we merge a PR from any branch into main, it will typically require only a single commit that very likely will have no errors, is yet another step that I am happy to avoid, as long as we can avoid problems.   I believe, with small teams, this is common practice, and it works fine as long as there are sufficient tests to ensure PRs won't cause unexpected problems.   I don't know what the dividing line between small and large teams is, but it probably depends a lot on the nature of the project as well, and how the work is typically distributed.

The Bikeshed repo already does have a moderately large test suite that is applied after each commit.  Running these same tests locally, before commit, is also possible, but not essential to ensure a reasonable level of integrity.   Sharing intermediate state of a draft PR with co-workers by way of committing changes that might still be very incomplete and full of errors is a very useful technique which we should not sacrifice. 

Whether origin/main itself with all the current accumulation of PRs has errors is another question.  I would expect that we should never merge in PRs that result in known errors, unless we have high confidence that they are spurious, or that we will quickly fix the errors in a subsequent PR.   Then the next boundary condition happens when origin/main is released to the public, which should be only happen once more rigorous testing and sufficient reflection results in even higher confidence that everything should be OK.  

So, is the discussion about whether PRs to origin/main should ever result in known errors, or whether we have sufficient tests to have confidence about whether known errors are relevant, or something else?

**Comment by tabatkins** [comment]
@domenic I'm pretty sad to see this level of condescension, yes. I asked specific questions and gave specific reasoning for my decisions, and I'm getting back generic platitudes about "good software engineering", implying strongly that I am being a *bad* software engineer by not immediately adopting your suggestions. 

Bureaucracy is not a bad word: it is both an essential part of any organization as it grows (to linearize communication pathways as the network of participants grows), *and* a significant cost to all participants (as they have to service the bureaucracy rather than working directly on the task or with each other). Adding bureaucracy needs justification.

> Whether origin/main itself with all the current accumulation of PRs has errors is another question. I would expect that we should never merge in PRs that result in known errors, unless we have high confidence that they are spurious, or that we will quickly fix the errors in a subsequent PR. Then the next boundary condition happens when origin/main is released to the public, which should only happen once more rigorous testing and sufficient reflection result in even higher confidence that everything should be OK.

Right. Bikeshed started as a single-developer project (with frequent but relatively minor contributions from others) with a small community of users (the CSSWG). At that time, working directly in main, and *serving* directly from main, made sense; the community was small and close-knit, so problems could be immediately communicated and fixed. Keeping main in a working condition was important, but short-term failures, quickly detected and fixed, were okay. Branches were used when doing larger projects that needed the checkpointing of commits, but might leave the codebase in a broken state until it's finished.

Bikeshed has grown since then. Currently the developer situation is still fairly similar - it's mostly just me, with occasional contributions from others. (In fact, spot contributions have mostly dried up as it's become larger and more complex.) So working in main, with occasional branches when necessary, is still pretty reasonable.

But the community has massively expanded, to encompass most of the W3C *and* a large and diverse set of disparate users elsewhere. I can no longer count on people reporting issues to me; it's more likely that they work around the issues instead. (I do still get plenty of bug reports, which I of course love.) And due to this size, it's more likely that every exposed version is used by a non-trivial number of people, so short-lived failures can still cause significant pain. Thus moving to a release model made a lot of sense, and this has been largely successful. This also means that users are *insulated* from whatever happens before the point of "cutting a release".

If Bikeshed ever starts getting more community contributions, more stringently maintaining a known-good `main` would make sense, so short-lived failures accidentally introduced by one dev don't overly affect the other devs. Until that happens (and I'm not confident it ever will), the current `main` shepherding continues to work fine. We could also change it if we believe that the current `main` shepherding is *preventing* people from feeling comfortable contributing, but I don't believe this is the case currently.

**Comment by tabatkins** [comment]
Version 3.11.22
===============

* Significant refactoring of the "dfn panels" that show up on definitions and list all the links in the document. To reduce the amount of repetitive HTML sent down the wire, these are now sent in a reduced JSON form and constructed in JS on load.

* Added a new `Ignore MDN Failures` metadata, allowing you to silence an MDN failure (to deal with transient errors due to something moving between specs and MDN temporarily being wrong about where a feature lives).

* You can use `<wpt skip-existence-check>` to make the WPT block, well, skip the existence check. This lets you use URLs that aren't in WPT at all, or aren't there *yet*, and still get a similar-looking test block. (No pass/fail data or the helper links like source vs live, but otherwise same as normal.)

Version 3.11.23
===============

* Fixed a bug in highlighting, where sometimes a `<pre class=include-code>` would drop the last line of the include. (#2521)

* Slightly improved how I auto-generate IDs for some links (that need to be linked to by dfn panels, etc.) If you were depending on those auto-genned IDs, uh, they're different now. (#2531)

Version 3.12.0
==============

* Refactored all of my darkmode handling to work properly. There's a new `Darkmode` metadata you can use to turn off Bikeshed's darkmode styles.

Version 3.12.1
==============

* Removed some final remnants of the old CSSWG testsuite data. If you don't know what that is, don't worry it didn't affect you.

Version 3.13.0
==============

* Removed the ABSTRACTATTR macro, now that nobody consumes it. (It was the same as the `ABSTRACT` macro, but escaped to be suitable for use in an attribute value. But `ABSTRACT` is HTML content so it was weird to do that in the first place.) This almost certainly doesn't affect you either.

* Removed flake8 from my linting pipeline, added ruff to replace it, and turned on a FUCKTON of lints. It can't *quite* replace pylint yet, but it's close, and it's SO FAST.

* If you do something that makes an invalid spec (like invoking Bikeshed without an input file, with no `.bs` file in the current directory), I now properly catch that and fail gracefully from several commands rather than dying with a Python stacktrace.

* Did a bunch of refactoring of [the Install/Use section of the docs](https://speced.github.io/bikeshed) to emphasize using `pipx` to install and manage Bikeshed, as a much easier and safer install tool, and to expand some examples of how to use Bikeshed within GitHub CI.

* Bikeshed now checks if your datafiles are relatively old and automatically attempts to update them, so you don't need to run `bikeshed update` yourself. I did a lot of refactoring in this code and it *seems* to be great, but please let me know if there are any issues.

**Comment by tabatkins** [comment]
Version 3.14.0
============

Let's try that HTML parser rewrite again! Coming at it more gradually now; I do one early pass over the document with the new parser and then just reserialize back to a string. I haven't yet replaced the datablock, markdown, or shorthand parsing, and I still pass it to the pre-existing HTML parser to actually build a tree, so mostly this is just some extra work being done, but those'll all be absorbed into this work in later phases.

As I say in <https://github.com/speced/bikeshed/pull/2602>, this brings a few benefits immediately:

* Detects some markup mistakes, like duplicated attributes.
* Correctly adds line numbers to everything in your source. (Some elements coming from metadata or datablocks still won't have correct line numbers.)
* Handles markdown code spans, code blocks, CSS `<<type>>` autolinks, and CSS `''maybe''` autolinks correctly, in the parser, rather than hackily via a regex.

It then still runs the datablock, markdown, and finally the existing HTML parser over the spec, so this is probably slightly slower at the moment, but those will be eaten by the parser in later phases.

Known issues with existing specs' markup:

If you were working around the hacky `` ` `` parsing by using `` \` `` inside of a code block
(which shouldn't have any parsing done inside)
it no longer does any parsing inside.
Fix: Remove the spare `\` so you just have valid JS again.
If you weren't working around this your spec was probably broken,
and now it's fixed!

The HTML parser runs before the Markdown parser right now
(fully integrating the two together is the next project),
so a tag broken across a line inside a blockquote
will parse incorrectly
(it's closed prematurely by the blockquote's `>` at the start of the next line)
Fix: just put the whole tag on one line for now.

Previously, `''&lt;foo>''` would make a maybe autolink to `<foo>` as a value.
Now it's equivalent to `<css>&amp;lt;foo></css>`,
which is broken,
but arguably it was always broken in the first place.
Fix: change to `''<foo>''`.

The Markdown behavior of "if you have spaces at both the start and end of a code span, remove one from each side" is now properly implemented. A few specs relied on it removing *any amount* of *any whitespace*, so now there's an extra space sometimes if you did a linebreak inside your code span for some reason.
Fix: Put the code span all on one line, or at least don't linebreak between the end of the content and the closing ticks.
(The serializer still collapses starting/ending whitespace down to a single character to make the content look better, so that might still be doing more stripping than you want too.)

**Comment by tabatkins** [comment]
A few small updates in the last few weeks.

Version 3.14.1
===========

* The AB is now a W3C group (I really need to move this data to bikeshed-data rather than in the Python...)
* (temporarily) allowed CR as whitespace, to fix a confusing parse error that some people were getting.

Version 3.14.2
============

* Made `=` in unquoted attribute values not an HTML parse error, so you can still have unquoted URLs with query strings in your attributes. (This'll be partially reverted in a future version, as the parse error apparently legitimately catches a lot of problems in the wild where people accidentally omit the space between two unquoted attributes, but I'll keep the exception for attributes known to contain URLs, like `href` and `src`.)

Version 3.14.3
============

* reverted the CR change from 3.14.1, instead properly normalized linebreaks when the source file comes from a URL. Turns out `requests` doesn't do this for you, while the rest of Python file loading does.

Version 3.14.4
============

* Added a `Die On` metadata, paralleling the `--die-on` command-line flag, that controls what level of error message results in the spec failing to build entirely. No change to existing documents (it still defaults to "fatal"), and the command-line flag still overrides anything coming from metadata.
* Allowed W3C Community Groups to use the `w3c/UD` status, since it's an unofficial status anyway, and useful to distinguish from their "real" draft statuses.

**Comment by tabatkins** [comment]
Version 3.14.5
============

* I now finally track *console output* as part of the testsuite as well. Previously I just monitored the actual spec output, but new fatal errors could show up without changing the output (and even changes to warnings could break builds that were using `--die-on=warning` to ensure a clean build). Shouldn't run into surprises like that anymore.
* When using `<pre class=simpledef>`, I no longer automatically sentence-case the keys, so you can do things like `{{someMethod()}}: describing the method` and the link will no longer get broken.

**Comment by tabatkins** [comment]
Version 4.0.0
=============

Woo, major version bump! Content-wise this isn't a break-worthy update, but Bikeshed has officially deprecated support for Python 3.7 (and 3.8), and now requires a minimum version of 3.9. As far as I can tell, this is the highest widely-supported version among default installs now; the old OSes that bundled 3.7 by default are all now out of long-term support.

* The link-checking lint now properly handles a timeout.
* If you're building a spec from within a github repo, the commit is now recorded in `<meta name=revision>`, to accord with the registered meta name for this purpose.
* Some minor fixed to how dfn panels work.
	* Clicking the "self-link" in the panel no longer triggers it to move down to the lower left corner; that only happens if you click one of the references to the dfn.
	* Dfn panels no longer get clipped by overflow on ancestors of the dfn. (This did require switching to positioning them more manually with JS, rather than being able to rely on CSS's "static position" concept; please let me know if anything looks wrong.)
	* Dfn panels now more intelligently avoid overflowing *the screen*, sliding nicely away from the edge to stay visible.
* In addition to *dfn* panels, we now have *link* panels, which show on hover and fade away more eagerly. They give a few useful bits of information about the link now, and will be developed more in the future.
* Added a new `--die-when` command-line argument, letting you toggle between a fatal error stopping Bikeshed *immediately*, or just stopping it at the end from actually producing output. Defaults to "late" now (continues processing, only dies at the end), so you can see all your fatal errors at once. You can pass `--die-when=early` if you have too many errors and want to fix them one-by-one.
* Both `--die-on` and `--die-when` can now be specified in your metadata (including in your group's defaults), as `Die On` and `Die When`, taking the same values as the command-line args.
* The manifest for Bikeshed's data files is now written in [KDL](https://kdl.dev) rather than a bespoke format, so it should be more easily changeable in the future. (This shouldn't affect your authoring experience at all.)
* This existed before, but it's now documented: if, for whatever reason, you need to skip Bikeshed's auto-update check it does at startup to ensure its data files are up-to-date (for example, if you're on a plane and dont' want to wait for the network to time out every time), you can pass `--no-update` on the command line.
* The `[TITLE]` macro, used to pipe the document's title into the `<title>` element, now omits markup in your `Title` metadata and just contains text. (`<title>` is specially parsed in HTML, similar to `<xmp>`.) This means you don't necessarily need to write both `Title` and `H1` metadata if you want to put some markup into your document's `<h1>`. (But if you're using markup with semantic visuals, like `<sup>` or `<sub>`, you probably do still want to do something different there.) This ended up fixing accidental markup getting spewed into the document title for several specs in my testsuite.
* Syntax highlighting no longer strips whitespace from the start and end of the block. Spot-checking a few places where this caused a minor output difference, it doesn't look like it actually causes a visual difference. But this does fix a bug that a few specs hit when combining highlighting with existing markup that subtly corrupted the text.
* If you had `<pre highlight=foo><code>...</code></pre>`, previously it would get *double* highlighted. (Shouldn't have been visible, given the highlight styling most people are using, but it was silly verbose markup.) No more!
* Fixed a crash bug if you tried to use an HTML character reference that either expanded into two codepoints or a single astral codepoint. (Nobody hit that in my testsuite, but still.)

Most of my time was spent on making Bikeshed's new HTML parser more powerful and more correct (mostly in <https://github.com/speced/bikeshed/commit/5c6f7af73bbe345046023dc3c6ba4704044ff07a>). This *shouldn't* have any big effect on you (some caveats, below), but every step forward I make here makes me so much happier.

* *All* of the early text-munging I previously did before running markdown/html parsers is now moved into the HTML parser: macros, curly apostrophes, and line-ending em dashes. This means Bikeshed will no longer accidentally modify things inside of script/style/xmp elements, markdown code spans or blocks, or Bikeshed "datablocks" (the `<pre class=foo>` things that are specially handled).
	* A few specs in my testsuite had manually corrected for this, using Markdown escapes so their JS template strings weren't detected as Markdown code spans. These'll need to be removed; the `\` is now output as part of your spec. Sorry. (However, more specs *hadn't* corrected for it and just had broken output; they're now automatically fixed.)
	* Macros are now valid **only** (a) in text, or (b) in attribute contents. Nothing in the testsuite used them anywhere else, so this should be fine, but FYI.
	* Macros can now correctly handle having more macros in their content. I'll catch accidentally-recursive macros so you won't hang Bikeshed with infinite replacements, tho.
* Several metadata lines that are meant to contain HTML are now fed to Bikeshed's new HTML parser as well, so you'll catch errors there too.
* So are your boilerplate files (so the new macro parser can do its job). Again, this might catch some HTML errors in your boilerplate; [open a PR in speced/bikeshed-boilerplate](https://github.com/speced/bikeshed-boilerplate/) if 
* A *stupid* amount of fiddly work went into making line numbers in error messages more common and more useful. Line numbers from things outside your core document now give their context, so you won't be confused looking at line 20 of your spec when it's actually an error at line 20 of your `header.include` file, etc.
* I also fixed a bunch of off-by-one errors in line numbering. If you still see *any* errors that misattribute their line number, please let me know.
	* Exception: line numbers like "`Line ~20`" (with the tilde) still happen for the remaining linking shorthands that don't know their original line number; they're instead reporting the line number of the nearest construct (HTML element or Markdown block) that *does* know its line number. If *that's* incorrect, tho, let me know.
	* Exception: If an error like the previous bullet point occurs in *top-level raw text* that doesn't get wrapped in a Markdown paragraph, it might report the line number of the wrapper element coming from your `header.include`. This sucks, I know.
	* Both of these will be fixed in an upcoming version before EOY, when I migrate all the stuff that gets parsed *after* Markdown/HTML parsing into the bespoke parser, so they'll know what their actual source line number is.
* HTML comments are handled slightly differently internally now; this should generally be undetectable (or work better!) except for perhaps some trailing whitespace appearing where it wasn't before (but *was* present in your markup). Please let me know if something looks to have messed up, tho.

I did a *lot* of careful test review to make sure things are as correct as possible. Aside from the bits I listed above, and the possibility of some new markup errors being flagged,

**Comment by tabatkins** [comment]
Version 4.0.1
===========

* Tiny fix - I had some invalid CSS in a new feature that was causing validators to complain.

**Comment by tabatkins** [comment]
Version 4.0.2
===========

* Fixed some cases where an `<l>` element wasn't turned into a valid HTML element.

Version 4.0.3
===========

* Totally refactored my JS/CSS inclusion code. Just an internal thing, tho scripts/styles will move around a bit since they're now alphabetized by their boilerplate name, and the couple that host JSON data changed how they store it.
* Rewrote several of the JS scripts that were unintentionally spending a lot of computation at page load. For large specs like DOM, this changes ~2s of computation to ~100ms.

Neither of these should actually affect specs in a user-observable fashion; let me know if anything seems to be newly wrong.

**Comment by tabatkins** [comment]
Version 4.1.0
=============

Several miscellaneous fixes; I wanna get these out before I dive back into the parser.

* Now that dfnpanels contain linking guidance, they're useful on *all* dfns, so Bikeshed now generates them for all dfns. (Previously, since they only contained links to where they were referenced, they were only generated for dfns that had local references.)
* The pass/fail colors for MDN panels was accidentally still inline, so they've been pulled out into a color var so they can be darkmoded.
* If you write `<dfn><<length [0, 10px]>></dfn>` (defining any type shorthand with a range), Bikeshed will now auto-infer that it is actually a *value* definition, rather than a *type* definition. This might change some IDs, if your dfns weren't tagged before and you were relying on auto-detection.
* I now properly handle Python errors raised during metadata parsing (rather than just exploding).
* When you turn off (with the `Boilerplate` metadata) a script or style with an associated style/script, I now properly omit the associated stuff too, without you having to manually turn them off.
* Updated the [WebIDL Parser](https://github.com/plinss/widlparser) version, to capture some recent changes.
* Bikeshed's auto-update feature that keeps your datafiles up-to-date now does a fast check for internet access, so you're not stuck waiting for a timeout when you're, say, on an airplane.
* Also, auto-update will no longer fall back to manual update if the manifest update fails; it's supposed to be a fast best-effort thing, and manual updating is SLOWWWWWW.
* Also, if manual updating throws an error, I now properly handle that.
* The Big Text functionality invoked by `bikeshed source` is completely rewritten.
	* Most importantly, it now keeps the original comment text in the document, so you can just fiddle with it inline and re-run the command, rather than having to delete the generated comment and add back a new `<!-- Big Text: foo -->` comment.
	* Created a second, narrower font file, too, and switched to it by default. You can fit more letters on screen now!
	* Font files are now written in KDL, rather than a bespoke format.

**Comment by tabatkins** [comment]
Version 4.1.3
===========

* Sorted the WPT path fragments from the previous version, so the output is stable rather than reliant on set iteration order.
* Removed the code in dfn popups that prevented the page from scrolling when the target was already somewhere on screen, as it also prevented the hash from updating.
* Firefox has a buggy implementation of .offsetTop on some internal table elements, and that made dfn/link popups misalign when their anchor was inside of a table with a caption. Switched to using getBoundingClientRect() to calculate their positions instead, which consistently works everywhere.

**Comment by tabatkins** [comment]
Version 4.1.4
===========

* The `--die-on` and `--die-when` flags are now properly piped into the metadata infrastructure, so they'll win over less powerful `Die On`/`When` metadata as they were intended to.
* HTML comments in def blocks are now properly ignored, as they were before the parsing switchover when I was unsafely stripping them from everywhere.

**Comment by tabatkins** [comment]
Version 4.1.7 & 4.1.8
=================

(4.1.7 was an accidental no-op release.) Tiny update!

* Added `htmlwg` as a recognized W3C Group
* Updated `widlparser`

**Comment by tabatkins** [comment]
Whoops, left a few version off the changelist. They're all minor except the last.

Version 4.1.9
=============

* Just adding a new WG (SOLID CG) to the data files.

Version 4.1.10
==============

* Made the var-highlighting colors (the backgrounds that show on up `<var>`s inside of `<div algorithm>` blocks when you click the vars) darkmode-aware.

Version 4.1.11
==============

* Made the default tracking-vector image darkmode-aware.

Version 4.1.12
==============

* Just adding a new group (W3T) to the datafiles.

Version 4.2.0
=============

* Moved the data about what groups and statuses exist, which statuses a group is allowed to use, etc. over to [a datafile in the bikeshed-boilerplate repo](https://github.com/speced/bikeshed-boilerplate/blob/main/boilerplate/doctypes.kdl). Previously, updating this requiring rolling a new Bikeshed version; now it can be updated alongside new boilerplates being added! No more silly updates like 4.1.9 and 4.1.12.

* If you're missing some IDs that the MDN data expects to be present in your spec, the error message is a little better now.

* Fixed a weird corner case: as long as there's only a single overload for the method, you can always autolink a method without passing any args: `{{Foo/bar()}}` links to the `bar()` method, regardless of how many args `bar()` actually takes.

	Because IDL methods and attributes occupy the same namespace, you were *also* allowed to autolink to methods without using parens at all, like `{{Foo/bar}}` to link to the `Foo.bar()` method. However, due to some slight mismatches in how two different systems interfaced, this only worked *if* the method could be called without arguments. If there *were* required arguments, it didn't work; but `{{Foo/bar()}}` (parens, no args) still worked fine.

	This is fixed now, tho there *are* a few specs in the end-to-end testing database which now have ambiguous autolinks as a result. It's likely a few specs in the wild might run into the problem as well.

* Fixed a crash bug that could occur while I was figuring out your spec's indent level, if the spec contained at least 50 space-indented lines but they were all indented with only a single space.

**Comment by tabatkins** [comment]
Version 4.2.1
=============

* Fixed a couple of W3C URLs <https://github.com/speced/bikeshed/commit/61fea1be4463fd1ad9182d767287ecf7d6044fc6>

* Made `bikeshed spec` pay attention to `--print=json` (previously, only `bikeshed refs` used it), and document that value. <https://github.com/speced/bikeshed/commit/296e6fd70c6cb8c8a7c774cc980601c21b793e45>

**Comment by tabatkins** [comment]
Version 4.2.2
=============

* Fixes the parser to allow for `<<foo [<<bar>>]>>` shorthands to be parsed,
	now that that's an allowed syntax for CSS grammars.

Version 4.2.3
=============

* Allow `<<foo []>>` (with an empty arg).

Version 4.2.4
=============

* Didn't realize that argparse (the cli library I use) doesn't correctly handle the same option name being used both at the top-level and in a subparser, so moved the `--no-update` top-level option to a new internal name. (No effect on user-facing code, except that `bikeshed --no-update spec` and `bikeshed update --skip-manifest` both correctly work now`.)
* Fixed #861, so refs with "significant" casing get that casing preserved when they're printed in indexes/etc. (Before, "dfn" type refs were always lowercased, so for example "ECMAScript" would print as "ecmascript" in the indexes.) There's some minor heuristics involved so sentence-starting refs don't preserve their casing accidentally.

Version 4.2.5 - 4.2.7
=====================

Insignificant version bumps, just me trying to get py.typed included, so mypy knows about Bikeshed's types.

Version 4.2.8
=============

* Bumped a lot of dependencies.
* Finally fixed the NodeJS favicon used by the CanIUse panels.

**Comment by tabatkins** [comment]
Version 5.0.0
=============

Big version bump this time, because it's a big update. This release includes the ~200 commits making up the new Big Parser Upgrade, which moves *all* of Bikeshed's "inline" special syntax (macros, link shorthand syntax like `[=...=]`, inline Markdown) into the bespoke HTML parser first introduced back in [Version 3.14](https://github.com/speced/bikeshed/issues/1773#issuecomment-1631507878).

Before I get into the implications of that, there's a few other misc fixes that got into this release:

* <https://github.com/speced/bikeshed/commit/ab9fa96ed68e84cf1e1bb83f528007c8629c9445> Tweaked the url for Process links in W3C specs.
* <https://github.com/speced/bikeshed/commit/2623e7a1a92d4bd0820d717d9ffe1ac5266834d9> Bumped the supported WPT manifest version to 9 (it was updated back at the end of November for unimportant reasons; nothing actually changed about the file)
* <https://github.com/speced/bikeshed/commit/138406a1ad69d16158ab926c1eb3d7ad07214647> Simplified the way I do indent inference to make it more correct, and removed the error message about your document being too short for me to infer an indentation size.
* Cleaned up some boilerplate/etc aspects of specs produced by the TAG.

Parser Updates
--------------

As I said, back in July 2023 I released version 3.14, with a new hand-written HTML parser. (There were some preceding releases that tried to use an earlier version of this parser, but 3.14 was the big one.) At the time the main effect of this was that I finally got accurate source-location information for all tags, something which was sorely lacking from my existing HTML parser.

This update expands the parser to handle *all inline Bikeshed syntax*. (That is, not block-level Markdown stuff like lists/etc.) Previously, these were handled by several distinct passes over the document at different times: 

* macros (`[FOO]`) were done as a regex over the original source
* then `<<foo>>` and `''foo''` autolinks - the former because they wouldn't survive a normal HTML parse, the second because they often contained additional markup, and that broke things in later passes
* then a block Markdown pass (handling paragraphs, lists, headings, etc)
* then an HTML parsing library to build the document
* then a pass over the text nodes of the document to find all the remaining inline Bikeshed shorthands (autolinks, inline markdown, etc). Each text node was processed for each type of shorthand in an arbitrary order, and the method I used could only find shorthands that lived *entirely* in a single text node, so any markup in the shorthand would prevent it from being recognized, and nesting shorthands only *sometimes* worked, if it just so happened that the outer shorthand was processed first, leaving the interior as a whole text node for the second to be recognized in.

As a result of these happening in several phases, escaping was complicated and inconsistent, depending on whether they were processed before or after the HTML parser ran. For example, `&#x5b;FOO]` would successfully escape a macro, keeping it as a plain `[FOO]` text in your document (because the macro parser ran before the HTML parser), but `&#x5b;=foo=]` would *not* escape a dfn autolink; the HTML parser would canonicalize the escape back into a `[` before the shorthand parser got a chance to run.

Also, because these were done via regexes, I didn't really have source location information anymore. The best I could do was *sometimes* report the line of the nearest element created by the Markdown block parser, which *did* know exactly what line of the source everything it created was from.

Now, the bespoke parser handles all of these consistently, at the same time. I know the precise source location of every inline construct; escapes are handled in a consistent manner for everything; etc.

As part of this, I also needed to start tracking the open elements, so I know whether I'm in an element that should suppress shorthands or not. This means I'm suddenly able to detect errors in tag nesting, missing end tags, or misspelled end tags, which seem to be reasonably common in practice.

Implications
------------

* If you were previously using `\[` to escape macros in `<script>` or similar, you no longer need to (and the `\` will now be left in your document, probably causing an error).
* Certain markdown links look like macros (like `[FOO](/foo)`). You could previously write `\[FOO](/foo)` to escape it at the macro-processing time, which would remove the `\`, so at inline-markdown processing time it saw just the markdown link markup. This is no longer necessary, since both are handled at the same time - `[FOO](/foo)` will be recognized as a markdown link properly, *not* a macro. (And the escaped version will produce *an escaped markdown link* now, aka plain text with []s.)
* Just kinda generally, if escaping previously looked like it was working weird, it should work "correctly" now, and if you used weird escapes it'll be a little broken. Sorry about that.

If you see any problems, please let me know. The code has been live for about two weeks so far and it seems like it's generally been a smooth transition.


Version 5.0.1
=============

* perf fixes, mostly noticeable in large specs
* 5.0.0 broke use of `|arg|: description` lines in argumentdef blocks. This is now manually handled.

5.0.1 is largely a perf release. I plan to try and get HTML converted to Bikeshed later this year, and needed to make sure that it could handle a document of that size. To do this, I just made a spec that was 15 copies of the DOM spec appended to itself, approximately the same filesize as the HTML source.

First try, with the 5.0.0 code, took 7s to process the original DOM source, and *770 seconds* (13min and change) to do the 15-doms spec. That was ridiculous, clearly I had some quadratic behavior.

After a bunch of honestly pretty easy fixes, 15-doms is down to 70s, and DOM itself is down to 5s. I have some evidence that the fixes might have slightly slowed down *small* specs (the testsuite seems to be a few seconds slower), but they're already fast enough that it seems insignificant.

**Comment by tabatkins** [comment]
Version 5.0.3
===========
* the `<l>` element (custom to Bikeshed) is now handled in the parser, guaranteeing that it parses shorthands *exactly* like normal. This also means it handles the CSS `''foo''` shorthand, which previously it was unable to do.
* Macros inside of the "data" portion of autolinks, like `[=[FOO]=]` previously worked by accident; in 5.0 I made them work on purpose when I realized it was necessary. Now it's off by default, to stick to my guns of "absolutely no parsing happens inside the data portion"; you can turn it on with `Markup Shorthands: macros-in-autolinks yes`.
* Added a handful of Bikeshed-specific HTML character references: ones for all of the ASCII punctuation, like `&bs[;` for the `[` character, and then a few for Infra purposes, `&bs<<;`, `&bs>>;`, and `&bs->;` to generate the curly braces and arrow mandated for the list/map/etc literal syntax.


Version 5.0.4
===========

* Made the timeout for `bikeshed echidna` configurable, and caught the timeout error and gave it a good error message (rather than vomiting a pretty large stack trace).
* Fiddled with widlparser's error messages so they report the true line number (rather than the block-local error number, with no indication of where the block is).

**Comment by tabatkins** [comment]
Version 5.0.5
===========

* Made the link-data section of autolinks only allow themselves to be spread over two lines. (Before, it just went until it found what looked like an closing sequence, which was possibly *many* lines further down... That produced confusing/unreadable error messages, and could break a lot more of the document inadvertently.)

**Comment by tabatkins** [comment]
Version 5.0.6
==========

* Updated `widlparser` dependency to 1.2, which fixes some bugs surrounding whether dictionary args in IDL blocks need to be optional or not.

**Comment by tabatkins** [comment]
Version 5.1.0
============

* Previously, if you were using a `<pre class=anchors>` block and had several anchors pointing to the same URL (for example, to a PDF, or to an IETF spec's section heading), the ref-hint popup would assume they were all the same ref and suggest the first one was the "canonical" spelling. No longer; it now correctly tracks anchor-block refs.
* Added `Link Checker Timeout` metadata, to control the timeouts for the `Complain About: broken-links on` option. Takes two integers, one for the timeout on individual links and one for the overall timeout. (The defaults remain `5` and `10`.)
* Added a new default-on `Markup Shorthands` value, `repository-links`. This allows syntax like `<#5>` and `<org/repo#5>` to expand into links to the appropriate GitHub issue. (The first relies on the standard repo detection - either your `Repository` metadata, or the git repo your source file is in.)


Version 5.1.1
===========
* Caught more variations of `widlparser` errors so they generate good line numbers again.
* Bikeshed's autolink conjugator now understands `(s)` suffix, like `group(s)`.
* `bikeshed template` can now specify what kind of template you want. The `bikeshed template base` is now the default, while `bikeshed template minimal` gives a template for a spec that turns off almost all the extra boilerplate stuff.
* Added a `bikeshed outline` command to quickly generate a document outline, with line numbers and IDs prominent.
* Tightened the behavior around `algorithm` on headings. The implicit algorithm now only applies to vars that are directly in that heading's section; vars in sub-sections don't see the algorithm. (This was mostly a perf issue; if you had a lot of vars that *weren't* in `<div algorithm>` blocks, it could spend a *lot* of time repeatedly seeking up to the root looking for algorithms.)
* Fixed a crash bug in `bikeshed update` from the CSSWG biblio file finally being removed.

**Comment by tabatkins** [comment]
Version 5.1.2
===========

Minor update, just a few bugfixes:
* Tab/space linting uses the normal line-numbering machinery for consistency, rather than manually including the line number in its message.
* Using `[LegacyFactoryFunction]` in WebIDL (used by HTML, hopefully not much else) is now allowed, with proper linking and no error message.
* The extra anchor elements added when you use `oldids` attribute are now *prepended* to the element's contents, rather than *appended*, so using it on longer elements will now properly scroll to the start of the element rather than the end.

**Comment by tabatkins** [comment]
Version 5.3.0
============

* Previously, link errors would give warnings in the console, but silently turn your link into a `span` in the output document. This was confusing sometimes, as you'd have what looked like some plain text with no indication that there was a failure. *Ideally* people fix all their link errors, of course, but authors aren't perfect.

	Now, the failed link is turned into a `<u class=link-error title="error message here">`, so it's more obvious something has gone wrong, without being *too* obnoxious about it.
* As part of that refactoring, I was able to track what links were in the running before they got eliminated by some filter, so I can use that in the error message. Right now, just the "no exported links" message uses it, as the most annoying one, to tell you what specs aren't exporting the term you want.

**Comment by tabatkins** [comment]
Version 5.3.1
==========

* Whoops, didn't lint my changes when I rebased earlier, so technically there was a crasher in the cddl handler code. Fixed.

**Comment by tabatkins** [comment]
Version 5.3.2
===========

Jeez, okay, *now* we're good. This is what happens when I get lazy and violate my self-imposed prereqs for cutting a release - things slip thru.

Quick post-mortem: the version of Bikeshed on my work desktop was a little out of date, so I happened to write the 5.3.0 patch against version 5.1.1 code. Linted, tested, everything was good, so I rebased onto 5.2.1 and cut the release.

Mistake: mypy was complaining. Quickly fixed it, pushed a new version. I also made a little fix in the nearby cddl code, because something looked a little messy and I was there already. Should have just been a one-line refactor.

Mistake: my change actually broke a cddl test. I'm *still* not sure how the code originally worked, but somehow it did, and now it didn't. Tracked this down, fixed it. Made certain lints and test suite were happy this time.

Also, part of the 5.3.0 code was non-deterministic (it serialized a `set`, which has undefined order and can change between machines). Looked fine in my tests, but got output in a slightly different order in CI; not a problem for users but breaks CI and needs fixing. Trivial fix, but I should have let it get caught by CI *before* I pushed 5.3.0 in the first place.

Because all of this *was* a tiny issue and was solveable as I was heading home, I left it in and just patched it. Had it been a larger or more confusing issue, I would have reverted 5.3.0 entirely, as I pledged to do last time I released a broken version. Not necessary this time, tho.

Anyway, that just reinforces why I have the "make sure the CI is green before you cut a release" rule in the first place. Luckily it triggered on a tiny non-issue. I'll be better about this in the future.

**Comment by tabatkins** [comment]
Version 5.3.3
==========

* The rules for dfn type inference weren't actually tested, but the inference is non-trivial, so I added some tests to make sure they worked right. Also added some error messages to catch some inference situations that are probably mistakes (trying to dfn something like `<foo()>`, which gets incorrectly inferred to be a `type`, not a `function`).
* Fixed #1443, where a local `biblio.json` was looked up in the current working directory rather than in the spec's directory, how all other locals are handled.
* Updated a bunch of dependencies.
* Fixed #3079, where using angle brackets in the "data" portion of an autolink, and no "visible text" portion, would format those brackets incorrectly; `[=parse a <syntax>=]` would output as `<a ...>parse a <syntax></a>`, which is obviously wrong.

**Comment by tabatkins** [comment]
Version 5.3.4
===========

* Support the new `STATEMENT` status for W3C specs (Currently, a new status needs a macro set in Bikeshed's code. I should generalize this...)

**Comment by tabatkins** [comment]
# Version 5.3.6

* Messed up the JSON output; fixed now.

# Version 5.4.0

* Made the block-Markdown pass optional, controlled by `Markup Shorthands: block-markdown off`. (Requested by HTML to shave a little time off, since they don't use any block Markdown constructs yet.)

**Comment by tabatkins** [comment]
Version 6.0.0
===========

New major version bump, because I shipped another major update to the parser!

In short, I replaced the `html5lib` library, which I used to do the "html source -> LXML tree" parsing step, with a hand-rolled tiny HTML parser that builds the tree directly using LXML SAX events. Since I was implicitly relying on html5lib to fix up a *lot* of little issues with the source text, I had to reproduce those fixes myself. In the course of doing that, I cleaned up a *lot* of technical debt thad had accumulated in the early parsing stages over the years.

This is *mostly* an under-the-covers change, and I've worked *very* hard to make sure that it results in as minimal a diff as possible for specs. Generally, you should just notice that a few more things get good line numbers in error messages, a few more markup errors get detected and reported, and Bikeshed itself is a bit faster (30-40% on larger specs!).

There are a few specs in my testing folder that had invalid markup that'll get interpreted in a slightly different way now, but in every case I saw, it's accompanied by error messages highlighting the problem now (where previously html5lib just silently fixed the markup in some arbitrary way).

If you notice any changes that you think are a mistake, please let me know ASAP.

**Comment by tabatkins** [comment]
Version 7.0.1 and 7.0.2
===================

* A two-parter fixing a crasher bug (7.0.1 did a quick and dirty fix that prevented the crash; 7.0.2 fixed it more thoroughly and got lint/tests fixed up). The bug involved attributes containing newlines, which could get accidentally rewritten by the markdown parser. Completely by accident, that previously didn't cause a problem, but I did some innocuous markup changes that happened to now cause it all to misparse in the final html->tree parse. Anyway, all multiline attributes now pass through the markdown parser without being screwed up.
* More generally, if I screw something up in the future around attributes, it'll get caught nicely and print a good error message, rather than crashing bikeshed.

Version 7.0.3
==========

* When parsing markdown links (like `[foo](http://example.com)`), if there was an escaped character in the destination it would get double-escaped and produce a broken link. Fixed.

**Comment by tabatkins** [comment]
Version 7.0.5
=============

* Just some fiddling with the internals of the `bikeshed template` command. You can now specify an outfile for them, rather than always printing to stdout!

Version 7.0.6
=============

* Caught some network errors that can happen while fetching inline issues.

Version 7.0.7
=============

* Never unlink biblio/section links, even when they're repeated in a paragraph. <https://github.com/speced/bikeshed/commit/65ddfc6f39f82f84e62bccf81368f4b66872c612>


Version 7.0.9
=============

A number of minor fixes, plus a bunch of deps bumps.

(7.0.8 was fat-fingered and missed a commit I'd merged in the web UI.)

* Removed some early checks that make sure you aren't using Python 2 (no longer necessary, come on) and aren't using a manually-installed version with out-of-date reqs (super rare use-case, and the testing requires tools that might not exist on non-manual installs). <https://github.com/speced/bikeshed/commit/36b1fb5a611e136ad1f3ba2882330d684f4f4a90>
* `iterable`/`maplike`/`setlike` now autolink in WebIDL blocks. <https://github.com/speced/bikeshed/commit/2d7d495fc1e9cc3ef5adc637459df8b9f4548520>
* Rename "Informative References" section title to "Non-Normative References", as "informative" is a term-of-art that isn't clear to everybody. <https://github.com/speced/bikeshed/commit/4e83f84e2bd522c52505d87ca9bf884cae4d7add>
* Tables in figures now auto-numbers as "Table N"! <https://github.com/speced/bikeshed/commit/ca1ad23fd11ad3a354a7c3913b1cec2e841d6e40>
* Some docker fixes <https://github.com/speced/bikeshed/commit/af78451121beb11e092b2f5853b88265aa949d54> <https://github.com/speced/bikeshed/commit/754a595e00268239983e4c66326b56811f87c309>
* Actually load link-defaults data when using a non-directory input source (like from a URL), whoops. <https://github.com/speced/bikeshed/commit/11686460900549ac72eea12f9efdb27f5fd519f2>
* Darken some syntax-highlighting colors when printing, for readability. <https://github.com/speced/bikeshed/commit/569f5e6c9a1bf8661ea7e608c61aa02bede631b8>

**Comment by tabatkins** [comment]
Version 7.0.10
==============

* As part of the big parser switch of v6, I unwittingly moved the "are you mixing tab and space indents" check to *before* the parser rather than after. This lost the ability to ignore the contents of style, script, pre, etc, which Bikeshed previously did. Fixed now, so the lint won't complain as much in places it shouldn't! <https://github.com/speced/bikeshed/commit/ccdd6e83fad350d2ff86a3180b003157fcbfb99c>

**Comment by tabatkins** [comment]
Version 7.0.11
============

* Just a dep bump. Notably, `widlparser` 1.5.0, which supports `async_sequence<T>`.

**Comment by tabatkins** [comment]
# Version 7.0.12

* Fixed handling of `<wbr>`, which accidentally broke in the v7 parser upgrade. #3304
* Stopped printing the element itself as part of lint/link-failure error message.
	
	The immediate reason for this is that LXML changes how it serializes attributes in a future Python, which was causing test breakage. I need to replace the few spots where I still lean on LXML here with custom serialization, but until I do, just removing the offending bits from the tested output does the job.

	But this was due anyway - I'd originally printed the offending element to help you *find* the error, back in the days before Bikeshed knew what source lines things came from. I always know an element's source location now, tho, and print it as part of the error message, so it's no longer necessary. It'd become gradually messier and less useful anyway, as I've loaded up elements with more bs-specific attributes for internal usage over time, and those showed up in the printing.

**Comment by tabatkins** [comment]
# Version 7.0.13

* An unrelated #3304 fix. Pygments *sometimes* deletes a whitespace prefix from the highlighted text, which I have to re-add, but my detector code for that was failing when the Pygments highlighter happens to emit the whitespace across several tokens (which, for example, the HTML highlighter did). Now same-color tokens get merged before I do any other processing, which appears to always fix this issue now.

**Comment by tabatkins** [comment]
# Version 7.1.0

* Long-needed rearranging of the [boilerplates](https://github.com/speced/bikeshed-boilerplate) folder structure, to properly handle org-level fallback for groups.

# Version 7.1.1

* 7.1.0 forgot to bump the internal data version, so if you updated an existing install to 7.1.0 without running `bikeshed update`, your active data files wouldn't be replaced with the new folder structure and bikeshed would fail to find any of the boilerplates. Fixed now, plus a full data-file refresh because might as well.

**Comment by tabatkins** [comment]
Version 7.1.2
===========

* Updated the syntax-highlighting file to be up-to-date on known metadata names <https://github.com/speced/bikeshed/commit/1d63dffff1e765dd83a8335626a5ce2e32c7bbe1>
* Significantly improved handling of Markdown blockquotes (<https://github.com/speced/bikeshed/pull/3317>). Bikeshed now requires you indicate explicitly that you're using md blockquotes (or that you're not, if it *looks like* you might be), since the syntax is somewhat intrusive with HTML parsing and I don't want to misparse things silently. Importantly, tho, starting lines with `>` inside of code blocks, datablocks, script/style, etc no longer screws up.
* Fixed a small bug caused when a spec manually used the `heading` class, which Bikeshed also automatically uses to mark headings. <https://github.com/speced/bikeshed/commit/548342ddcf484a716c347dcf18586a5e19dbc0ff>

Version 7.1.3
===========

* WebIDL errors now show their line number (or at least, the line number of the `pre` opening tag)
* `argumentdef` blocks for methods with overloads defined now consistently work (phew, <https://github.com/speced/bikeshed/issues/1192> was an old bug!)
